File: security.jl
mutable struct SecurityRequirement <: Comparable
    name::Vector{String}
end

SecurityRequirement() = SecurityRequirement(Vector{String}())

function parse!(s::SecurityRequirement, data::OrderedDict{Any,Any})
    for (key, value) in data
        if key == "name"
            for v in value push!(s.name, v) end
        end
    end
end

function parse!(s::Vector{SecurityRequirement}, data::Vector{OrderedDict{Any,Any}})
    for d in data
        sr = SecurityRequirement()
        parse!(sr, d)
        push!(s, sr)
    end
end

mutable struct OAuthFlow <: Comparable
    authorizationUrl::URI
    tokenUrl::URI
    refreshUrl::URI
    scopes::OrderedDict{String,String}
end

OAuthFlow() = OAuthFlow(URI(), URI(), URI(), OrderedDict{String,String}())

function parse!(o::OAuthFlow, data::OrderedDict{Any,Any})
    for (key,value) in data
        if key == "authorizationUrl" o.authorizationUrl = URI(value) end
        if key == "tokenUrl" o.authorizationUrl = URI(value) end
        if key == "refreshUrl" o.refreshUrl = URI(value) end
        if key == "scopes"
             for (k,v) in value o.scopes[k] = v end
         end
    end
end

mutable struct OAuthFlows <: Comparable
    implicit::OAuthFlow
    password::OAuthFlow
    clientCredentials::OAuthFlow
    authorizationCode::OAuthFlow
end

OAuthFlows() = OAuthFlows(OAuthFlow(), OAuthFlow(), OAuthFlow(), OAuthFlow())

function parse!(o::OAuthFlows, data::OrderedDict{Any,Any})
    for (key,value) in data
        if key == "implicit" parse!(o.implicit, value) end
        if key == "password" parse!(o.password, value) end
        if key == "clientCredentials" parse!(o.clientCredentials, value) end
        if key == "authorizationCode" parse!(o.authorizationCode, value) end
    end
end

mutable struct SecurityScheme <: Comparable
    ref::String
    type::String
    description::String
    name::String
    in::String
    scheme::String
    bearerFormat::String
    flows::OAuthFlows
    openIdConnectUrl::URI
end

SecurityScheme() = SecurityScheme("", "", "", "", "", "", "", OAuthFlows(), URI())

function parse!(s::SecurityScheme, data::OrderedDict{Any,Any})
    for (key,value) in data
        if key == "\$ref" s.ref = value end
        if key == "type" s.type = value end
        if key == "description" s.description = value end
        if key == "name" s.name = value end
        if key == "in" s.in = value end
        if key == "scheme" s.scheme = value end
        if key == "bearerFormat" s.bearerFormat = value end
        if key == "flows" parse!(s.bearerFormat, value) end
        if key == "openIdConnectUrl" s.openIdConnectUrl = URI(value) end
    end
end

function parse!(s::Vector{SecurityRequirement}, data::Vector{OrderedDict{Any,Any}})
    for d in data
        sr = SecurityRequirement()
        parse!(sr, d)
        push!(s, sr)
    end
end

File: src/parser/security.cpp
//
// Created by Rakesh on 19/09/2025.
//

#include "parser.hpp"
#include "model/security.hpp"

#include <boost/url/parse.hpp>

template <>
void spt::parser::parse( model::SecurityRequirement& m, c4::yml::ConstNodeRef node )
{
  for ( const auto& value : node.children() )
  {
    auto vec = std::vector<std::string>{};
    vec.reserve( 8 );
    for ( const auto& v : value.children() )
    {
      auto& sv = vec.emplace_back();
      v >> sv;
    }
    auto key = std::string_view{ value.key() };
    m.values.try_emplace( std::string{ key }, std::move( vec ) );
  }
}

template <>
void spt::parser::parse( model::OAuthFlow& m, c4::yml::ConstNodeRef node )
{
  for ( const auto& child : node.children() )
  {
    if ( child.key() == "authorizationUrl" ) m.authorizationUrl = parse<boost::urls::url>( child );
    if ( child.key() == "tokenUrl" ) m.tokenUrl = parse<boost::urls::url>( child );
    if ( child.key() == "refreshUrl" ) m.refreshUrl = parse<boost::urls::url>( child );
    if ( child.key() == "scopes" )
    {
      for ( const auto& v : child.children() )
      {
        auto value = std::string{};
        v >> value;
        auto key = std::string_view{ v.key() };
        m.scopes.try_emplace( std::string{ key }, std::move( value ) );
      }
    }
  }
}

template <>
void spt::parser::parse( model::OAuthFlows& m, c4::yml::ConstNodeRef node )
{
  for ( const auto& child : node.children() )
  {
    if ( child.key() == "implicit" ) m.implicit = parse<model::OAuthFlow>( child );
    if ( child.key() == "password" ) m.password = parse<model::OAuthFlow>( child );
    if ( child.key() == "clientCredentials" ) m.clientCredentials = parse<model::OAuthFlow>( child );
    if ( child.key() == "authorizationCode" ) m.authorizationCode = parse<model::OAuthFlow>( child );
  }
}

template <>
void spt::parser::parse( model::SecurityScheme& m, c4::yml::ConstNodeRef node )
{
  for ( const auto& child : node.children() )
  {
    if ( child.key() == "openIdConnectUrl" ) m.openIdConnectUrl = parse<boost::urls::url>( child );
    if ( child.key() == "$ref" ) child >> m.ref;
    if ( child.key() == "type" ) child >> m.type;
    if ( child.key() == "description" ) child >> m.description;
    if ( child.key() == "name" ) child >> m.name;
    if ( child.key() == "in" ) child >> m.in;
    if ( child.key() == "scheme" ) child >> m.scheme;
    if ( child.key() == "bearerFormat" ) child >> m.bearerFormat;
    if ( child.key() == "flows" ) m.flows = parse<model::OAuthFlows>( child );
  }
}
